File: app/src/test/java/com/soundcloud/android/associations/FollowingOperationsTest.java
package com.soundcloud.android.associations;

import static com.soundcloud.android.associations.UpdateFollowingCommand.UpdateFollowingParams;
import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.when;

import com.soundcloud.android.events.EventQueue;
import com.soundcloud.android.events.FollowingStatusEvent;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.sync.NewSyncOperations;
import com.soundcloud.android.sync.SyncResult;
import com.soundcloud.android.sync.Syncable;
import com.soundcloud.android.testsupport.fixtures.ModelFixtures;
import com.soundcloud.android.users.User;
import com.soundcloud.android.users.UserAssociationStorage;
import com.soundcloud.android.users.UserItem;
import com.soundcloud.rx.eventbus.EventBusV2;
import com.soundcloud.rx.eventbus.TestEventBusV2;
import io.reactivex.Maybe;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.observers.TestObserver;
import io.reactivex.schedulers.Schedulers;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

@RunWith(MockitoJUnitRunner.class)
public class FollowingOperationsTest {

    private static final int FOLLOWER_COUNT = 2;
    private FollowingOperations operations;

    @Mock private NewSyncOperations syncOperations;
    @Mock private UpdateFollowingCommand updateFollowingCommand;
    @Mock private UserAssociationStorage userAssociationStorage;
    @Mock private EventBusV2 mockEventBus;
    @Captor private ArgumentCaptor<UpdateFollowingParams> commandParamsCaptor;
    @Captor private ArgumentCaptor<FollowingStatusEvent> followingChangedEventArgumentCaptor;

    private TestEventBusV2 eventBus = new TestEventBusV2();
    private Scheduler scheduler = Schedulers.trampoline();
    private Urn targetUrn = Urn.forUser(123);

    @Before
    public void setUp() throws Exception {
        operations = new FollowingOperations(eventBus,
                                             updateFollowingCommand,
                                             scheduler,
                                             userAssociationStorage,
                                             ModelFixtures.entityItemCreator(),
                                             syncOperations);

        when(updateFollowingCommand.toSingle(any(UpdateFollowingParams.class))).thenReturn(Single.just(FOLLOWER_COUNT));
    }

    @Test
    public void toggleFollowingStoresThenSyncThenEmitsChangeSet() {
        final FollowingOperations ops = new FollowingOperations(mockEventBus,
                                                                updateFollowingCommand,
                                                                scheduler,
                                                                userAssociationStorage,
                                                                ModelFixtures.entityItemCreator(),
                                                                syncOperations);

        when(syncOperations.failSafeSync(Syncable.MY_FOLLOWINGS)).thenReturn(Single.just(SyncResult.synced()));

        ops.toggleFollowing(targetUrn, true).test();

        final InOrder inOrder = Mockito.inOrder(updateFollowingCommand, syncOperations, mockEventBus);
        inOrder.verify(updateFollowingCommand).toSingle(commandParamsCaptor.capture());
        inOrder.verify(syncOperations).failSafeSync(Syncable.MY_FOLLOWINGS);
        inOrder.verify(mockEventBus).publish(eq(EventQueue.FOLLOWING_CHANGED), followingChangedEventArgumentCaptor.capture());

        final UpdateFollowingParams params = commandParamsCaptor.getValue();
        assertThat(params.following).isTrue();
        assertThat(params.targetUrn).isEqualTo(targetUrn);

        final FollowingStatusEvent event = followingChangedEventArgumentCaptor.getValue();
        assertThat(event).isEqualTo(FollowingStatusEvent.createFollowed(targetUrn, FOLLOWER_COUNT));
    }

    @Test
    public void onUserFollowedEmitsFollowedUser() {
        final TestObserver<UserItem> testObserver = operations.populatedOnUserFollowed().test();

        final FollowingStatusEvent event = FollowingStatusEvent.createFollowed(targetUrn, FOLLOWER_COUNT);
        final User following = ModelFixtures.user();
        when(userAssociationStorage.followedUser(targetUrn)).thenReturn(Maybe.just(following));

        eventBus.publish(EventQueue.FOLLOWING_CHANGED, event);

        testObserver.assertValues(ModelFixtures.userItem(following));
    }

    @Test
    public void onUserUnFollowedEmitsUrnOfUnfollowedUser() {
        final TestObserver<Urn> subscriber = operations.onUserUnfollowed().test();

        final FollowingStatusEvent event = FollowingStatusEvent.createUnfollowed(targetUrn, FOLLOWER_COUNT);
        eventBus.publish(EventQueue.FOLLOWING_CHANGED, event);

        subscriber.assertValues(targetUrn);
    }

}

File: app/src/test/java/com/soundcloud/android/associations/FollowingOperationsTest.kt
package com.soundcloud.android.associations

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import com.soundcloud.android.events.EventQueue
import com.soundcloud.android.events.FollowingStatusEvent
import com.soundcloud.android.sync.NewSyncOperations
import com.soundcloud.android.sync.SyncResult
import com.soundcloud.android.sync.Syncable
import com.soundcloud.android.testsupport.fixtures.ModelFixtures
import com.soundcloud.android.users.FollowingStorage
import com.soundcloud.android.users.User
import com.soundcloud.android.users.UserItemRepository
import com.soundcloud.android.users.UserRepository
import com.soundcloud.propeller.ChangeResult
import com.soundcloud.rx.eventbus.EventBusV2
import com.soundcloud.rx.eventbus.TestEventBusV2
import io.reactivex.Maybe
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import org.assertj.core.api.Assertions
import org.assertj.core.api.Java6Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

@Suppress("IllegalIdentifier")
@RunWith(MockitoJUnitRunner::class)
class FollowingOperationsTest {

    private val followerCount = 2

    private lateinit var operations: FollowingOperations

    @Mock private lateinit var syncOperations: NewSyncOperations
    @Mock private lateinit var mockEventBus: EventBusV2
    @Mock private lateinit var userRepository: UserRepository
    @Mock private lateinit var userItemRepository: UserItemRepository
    @Mock private lateinit var followingStorage: FollowingStorage
    @Captor private lateinit var followingChangedEventArgumentCaptor: ArgumentCaptor<FollowingStatusEvent>
    private var userInfoSubscribed: Boolean = false
    private var insertFollowingSubscribed: Boolean = false
    private var syncSubscribed: Boolean = false

    private val eventBus = TestEventBusV2()
    private val scheduler = Schedulers.trampoline()
    private val userBuilder = ModelFixtures.userBuilder()
    private val targetUrn = userBuilder.build().urn()

    @Before
    @Throws(Exception::class)
    fun setUp() {
        userInfoSubscribed = false
        insertFollowingSubscribed = false
        syncSubscribed = false
        operations = FollowingOperations(eventBus,
                                         scheduler,
                                         userRepository,
                                         userItemRepository,
                                         syncOperations,
                                         followingStorage)
    }

    @Test
    fun toggleFollowingStoresThenSyncThenEmitsChangeSet() {
        val ops = FollowingOperations(mockEventBus,
                                      scheduler,
                                      userRepository,
                                      userItemRepository,
                                      syncOperations,
                                      followingStorage)

        setupFollowings(userBuilder.build(), true, true)

        ops.toggleFollowing(targetUrn, true).test()

        val inOrder = Mockito.inOrder(syncOperations, mockEventBus)
        inOrder.verify(syncOperations).failSafeSync(Syncable.MY_FOLLOWINGS)
        inOrder.verify(mockEventBus).publish(ArgumentMatchers.eq(EventQueue.FOLLOWING_CHANGED), followingChangedEventArgumentCaptor.capture())

        val event = followingChangedEventArgumentCaptor.value
        assertThat(event).isEqualTo(FollowingStatusEvent.createFollowed(targetUrn, followerCount))
        assertThat(syncSubscribed).isTrue()
    }

    @Test
    fun onUserFollowedEmitsFollowedUser() {
        val testObserver = operations.populatedOnUserFollowed().test()

        val event = FollowingStatusEvent.createFollowed(targetUrn, followerCount)
        val following = ModelFixtures.userItem()
        whenever(userItemRepository.userItem(targetUrn)).thenReturn(Maybe.just(following))

        eventBus.publish(EventQueue.FOLLOWING_CHANGED, event)

        testObserver.assertValues(following)
    }

    @Test
    fun onUserUnFollowedEmitsUrnOfUnfollowedUser() {
        val subscriber = operations.onUserUnfollowed().test()

        val event = FollowingStatusEvent.createUnfollowed(targetUrn, followerCount)
        eventBus.publish(EventQueue.FOLLOWING_CHANGED, event)

        subscriber.assertValues(targetUrn)
    }

    @Test
    fun `updates followed user on unfollow`() {
        val currentFollowingState = true
        val newFollowingState = false
        val user = userBuilder.followersCount(1).build()

        setupFollowings(user, newFollowingState, currentFollowingState)

        val test = operations.toggleFollowing(user.urn(), newFollowingState).test()

        test.assertComplete()

        Assertions.assertThat(eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED).size).isEqualTo(1)
        (eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED)[0]).let {
            assertThat(it.urn()).isEqualTo(user.urn())
            assertThat(it.isFollowed).isFalse()
            assertThat(it.followingsCount()).isEqualTo(0)
        }
        verify(userRepository).updateFollowersCount(user.urn(), 0)
        verify(followingStorage).insertFollowing(user.urn(), newFollowingState)
        assertThat(userInfoSubscribed).isTrue()
        assertThat(insertFollowingSubscribed).isTrue()
    }

    @Test
    fun `updates not followed user on follow`() {
        val currentFollowingState = false
        val newFollowingState = true
        val user = userBuilder.followersCount(0).build()

        setupFollowings(user, newFollowingState, currentFollowingState)

        val test = operations.toggleFollowing(user.urn(), newFollowingState).test()

        test.assertComplete()

        Assertions.assertThat(eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED).size).isEqualTo(1)
        (eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED)[0]).let {
            assertThat(it.urn()).isEqualTo(user.urn())
            assertThat(it.isFollowed).isTrue()
            assertThat(it.followingsCount()).isEqualTo(1)
        }
        verify(userRepository).updateFollowersCount(user.urn(), 1)
        verify(followingStorage).insertFollowing(user.urn(), newFollowingState)
        assertThat(userInfoSubscribed).isTrue()
        assertThat(insertFollowingSubscribed).isTrue()
    }

    @Test
    fun `does not change user when trying to follow followed user`() {
        val currentFollowingState = true
        val newFollowingState = true
        val user = userBuilder.followersCount(1).build()

        setupFollowings(user, newFollowingState, currentFollowingState)

        val test = operations.toggleFollowing(user.urn(), newFollowingState).test()

        test.assertComplete()

        Assertions.assertThat(eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED).size).isEqualTo(1)
        (eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED)[0]).let {
            assertThat(it.urn()).isEqualTo(user.urn())
            assertThat(it.isFollowed).isTrue()
            assertThat(it.followingsCount()).isEqualTo(1)
        }
        verify(userRepository).updateFollowersCount(user.urn(), 1)
        verify(followingStorage).insertFollowing(user.urn(), newFollowingState)
        assertThat(userInfoSubscribed).isTrue()
        assertThat(insertFollowingSubscribed).isTrue()
    }

    @Test
    fun `does not change user when trying to unfollow not followed user`() {
        val currentFollowingState = false
        val newFollowingState = false
        val user = userBuilder.followersCount(0).build()

        setupFollowings(user, newFollowingState, currentFollowingState)

        val test = operations.toggleFollowing(user.urn(), newFollowingState).test()

        test.assertComplete()

        Assertions.assertThat(eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED).size).isEqualTo(1)
        (eventBus.eventsOn(EventQueue.FOLLOWING_CHANGED)[0]).let {
            assertThat(it.urn()).isEqualTo(user.urn())
            assertThat(it.isFollowed).isFalse()
            assertThat(it.followingsCount()).isEqualTo(0)
        }
        verify(userRepository).updateFollowersCount(user.urn(), 0)
        verify(followingStorage).insertFollowing(user.urn(), newFollowingState)
        assertThat(userInfoSubscribed).isTrue()
        assertThat(insertFollowingSubscribed).isTrue()
    }

    private fun setupFollowings(user: User, newFollowingState: Boolean, currentFollowingState: Boolean = false) {
        val userUrn = user.urn()
        whenever(userRepository.userInfo(userUrn)).thenReturn(Maybe.just(user).doOnSubscribe { userInfoSubscribed = true })

        whenever(followingStorage.isFollowing(userUrn)).thenReturn(Single.just(currentFollowingState))

        whenever(userRepository.updateFollowersCount(eq(userUrn), any())).thenReturn(Single.just(ChangeResult(1)))
        whenever(followingStorage.insertFollowing(userUrn, newFollowingState)).thenReturn(Single.just(1L).doOnSubscribe { insertFollowingSubscribed = true })

        whenever(syncOperations.failSafeSync(Syncable.MY_FOLLOWINGS)).thenReturn(Single.just(SyncResult.synced()).doOnSubscribe { syncSubscribed = true })
    }
}
